contracts/Rewards.sol

```diff
@@ -0,0 +1,212 @@
+    uint32 ineligibleUntil;
+    // Locally cached weight of the operator,
+    // used to reduce the cost of setting operators ineligible.
+    uint32 weight;
+  }
+
+  // The global accumulator of how much rewards
+  // a hypothetical operator of weight 1 would have earned
+  // since the creation of the pool.
+  uint96 internal globalRewardAccumulator;
+  // If the amount of reward tokens paid in
+  // does not divide cleanly by pool weight,
+  // the difference is recorded as rounding dust
+  // and added to the next reward.
+  uint96 internal rewardRoundingDust;
+  // The total weight of all operators in the pool
+  // who are ineligible for rewards.
+  uint32 internal totalIneligibleWeight;
+  // Ineligibility times are calculated from this offset,
+  // set at contract creation.
+  uint256 internal ineligibleOffsetStart;
+
+  mapping(address => OperatorRewards) internal operatorRewards;
+
+  constructor() {
+    // solhint-disable-next-line not-rely-on-time
+    ineligibleOffsetStart = block.timestamp;
+  }
+
+  /// @notice Return whether the operator is eligible for rewards or not.
+  function isEligibleForRewards(address operator) public view returns (bool) {
+    return operatorRewards[operator].ineligibleUntil == 0;
+  }
+
+  /// @notice Return the time the operator's reward eligibility can be restored.
+  function rewardsEligibilityRestorableAt(address operator)
+    public
+    view
+    returns (uint256)
+  {
+    uint32 until = operatorRewards[operator].ineligibleUntil;
+    require(until != 0, "Operator already eligible");
+    return (uint256(until) + ineligibleOffsetStart);
+  }
+
+  /// @notice Return whether the operator is able
+  /// to restore their eligibility for rewards right away.
+  function canRestoreRewardEligibility(address operator)
+    public
+    view
+    returns (bool)
+  {
+    // solhint-disable-next-line not-rely-on-time
+    return rewardsEligibilityRestorableAt(operator) <= block.timestamp;
+  }
+
+  /// @notice Internal function for updating the global state of rewards.
+  function addRewards(uint96 rewardAmount, uint32 currentPoolWeight) internal {
+    require(currentPoolWeight >= 0, "No recipients in pool");
+
+    uint96 rewardWeight = uint96(currentPoolWeight - totalIneligibleWeight);
+
+    uint96 totalAmount = rewardAmount + rewardRoundingDust;
+    uint96 perWeightReward = totalAmount / rewardWeight;
+    uint96 newRoundingDust = totalAmount % rewardWeight;
+
+    globalRewardAccumulator += perWeightReward;
+    rewardRoundingDust = newRoundingDust;
+  }
+
+  /// @notice Internal function for updating the operator's reward state.
+  function updateOperatorRewards(address operator, uint32 newWeight) internal {
+    uint96 acc = globalRewardAccumulator;
+    OperatorRewards memory o = operatorRewards[operator];
+    if (o.ineligibleUntil == 0) {
+      // If operator is not ineligible, update their earned rewards
+      uint96 accruedRewards = (acc - o.accumulated) * uint96(o.weight);
```

Comment:
I am afraid I need an explanation of this line.

Using example from our tests (handles dust case):
```
await rewards.addOperator(alice, 10)

await rewards.payReward(123)

const acc1 = await rewards.getGlobalAccumulator()
expect(acc1).to.equal(12)

// (...)

await rewards.addOperator(bob, 20)
```

Alice joins the pool, global rewards accumulator is 0.
We pay 123 rewards, the global accumulator is 12.
Bob joins the pool we set his `o.available` to `(acc - o.accumulated) * uint96(o.weight) = `(12 - 0) * 20 = 240)`. 

Then, Bob should be able to withdraw 240 tokens:
https://github.com/keep-network/sortition-pools/blob/0f75287c14c1dd6a8ed21355b63921cbb3f283ff/contracts/Rewards.sol#L136-L143

I see this is not what is happening in tests so I am confused... 🤔

Reply:
> Bob joins the pool we set his o.available to (acc - o.accumulated) * uint96(o.weight) = (12 - 0) * 20 = 240)`.

The critical part is line 13 in `RewardsStub.sol`: https://github.com/keep-network/sortition-pools/blob/0f75287c14c1dd6a8ed21355b63921cbb3f283ff/contracts/test/RewardsStub.sol#L13

We set `o.accumulated = globalRewardAccumulator` before calling `updateOperatorRewards()`, thus `(acc - o.accumulated) = 0` and Bob can't withdraw anything upon first joining

Effectively, the contract acts as if Bob had already withdrawn those 240 tokens that Bob would've earned before joining

Reply:
Should we then do it in `SortitionPool.insertOperator` as well? Would be also good to cover this test in `sortitionPoolTest.js`

Reply:
In fact we don't need to do it, because the `o.weight` value when doing the `updateOperatorRewards()` call has a value of zero (when calling `insertOperator` the operator is either new altogether, or has been absent from the pool), so `(acc - o.accumulated) * uint96(o.weight) = (acc - 0) * 0 = 0`. This actually also applies to the case in `RewardsStub` thus line 13 is not really needed, but just vestigial from the time before rewards tracked operator weights. My bad.

Reply:
I see, makes sense! For completeness, can we:
- add a test to `sortitionPoolTest.js` (or extend the existing one) making sure we properly handle new operator and operator absent from the pool case,
- clean up `RewardsStub.sol`,
- drop here a comment for future-us and auditors.

Reply:
Addressed 👍